### PR TITLE
Create a framework for testing actual output of the enginebuffer.

### PR DESCRIPTION
**Pull request by ywwg**
After a call to ProcessBuffer(), tests can now compare the contents
of the output buffer against a golden data file representing expected
output.  This is useful for catching regressions in the engine code
as well as seeing how changes in the engine code affect audio output.

Included in this commit is a script called AudioPlot.py which has
been very useful to me in the past for analyzing audio data. It
requires matplotlib to be installed.

### CODE REVIEW

File: src/engine/enginebuffer.cpp [hunk 1]
```diff
@@ -1130,6 +1132,15 @@ void EngineBuffer::process(CSAMPLE* pOutput, const int iBufferSize) {
     }
 #endif
 
+    if (m_pTestSaveBuffer != NULL) {
```
Comment (daschuer) [inline src/engine/enginebuffer.cpp]: I would prefer to add this block in a process() override function. 
This way it dos not pollute the code under test.

**Comment by daschuer** [comment]
Nice idea!

**Comment by daschuer** [comment]
An extra readme or a equivalent text at the top of AudioPlot.py would help users who have just found the script. 
By the way: Why it is camel case? IMHO it is a good practice to don't use caps is command line tools.

**Comment by ywwg** [comment]
I did some refactoring so that we can look at the master output buffer instead of the enginebuffer.  This will allow us to test way more of the engine, like effects.  The pattern is pretty clean, too.  I had to regenerate the golden data because the tracks had been set to play by default.

File: scripts/audioplot.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    plt.show()
+
+    return
+
+def usage():
+    print """
+AudioPlot.py:
+-c              pick column to plot (more than one is OK, 1-based indexes, or "all")
+-f              file to plot from
```
Comment (daschuer) [inline scripts/audioplot.py]: An example would be nice, that actually works in the path hierarchy.

**Comment by daschuer** [comment]
LGTM, just a remaining doku idea.
Thank you.

File: scripts/audioplot.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+#
+# Each column will be plotted as a separate curve on the same time series.
+# Matplotlib is used to display the result.
+
+
+from matplotlib import pyplot as plt
+from matplotlib.lines import Line2D
+import time, string, glob, sys
+
+import getopt
```
Comment (kain88-de) [inline scripts/audioplot.py]: The standard Python way to parse cli options is argparse. Removes a lot of the boilerplate code. I'll write you a PR to make this script nicer

Comment (kain88-de) [inline scripts/audioplot.py]: WTF I can't make a PR to your repository oO. Anyway this is the commit.

https://github.com/kain88-de/mixxx/commit/ee4024de4bdf702e56c2a20e7f92955dfd692c70

**Comment by kain88-de** [comment]
looks really nice. Can you add some comments how to produce the golden test data. Otherwise LGTM.

**Comment by ywwg** [comment]
> produce the golden test data

this is included in the documentation for the assertBufferMatchesGolden function.

File: scripts/audioplot.py (hunk 3 above)
Comment (ywwg) [inline scripts/audioplot.py]: > commit.

Thanks for the rewrite, the original audioplot was a late-night hack that never really got looked at again.  Unfortunately the version you linked to is broken.  I'll commit this framework as is, and if you can fix the issues with the rewrite you can submit a regular PR

**Pull request merged**